### PR TITLE
Enhancements of PowerConverters (incl. fix for Signal2mPulse and new Voltage2DutyCycle)

### PR DESCRIPTION
**Pull request by AHaumer**
Replaces PR #2955, summarizing the following enhancements:
- Implements bugfix for Signal2mPulse (PR #2965 for master and #2967 for 3.2.3)
- Enhanced Voltage2DutyCycle
- VoltageToDutyCycle kept for compatibility, marked as obsolete and moved to ObsoleteModelica4
- Adapts Machines.Examples.ControlledDCDrives.Utilities.SwitchingDcDc (new Voltage2DutyCycle)
- Enhanced Examples.DCDC.{ChopperStepUp, ChopperStepDown}:
  - add inductor and smoothing capacitor
  - change default parameterization (more realistic and comparable)
  - StepDown results in 60 V -> 12 V / 1.2 A, StepUp results in 60 V -> 75 V / 1.2 A
- Implement ACAC.Triac
- Implement ACAC.Control.{VoltageToAngle, SoftStartControl}
- Implement Examples.ACAC.SoftStarter

### CODE REVIEW

**Comment by AHaumer** [comment]
yes, but not #2967

**Comment by dietmarw** [comment]
I suggest to wait with this enhancement for the benefit of the renaming of electrical machines and multiphase --> polyphase. This will save you some conversion work later on.

**Comment by christiankral** [comment]
I didn't want to re-request a review -- it happened accidentally.

File: Modelica/Electrical/PowerConverters.mo [hunk 1]
```diff
@@ -90,6 +92,41 @@ is provided.
...
+    class ACACConcept "AC/AC converter concept"
+      extends Modelica.Icons.Information;
+      annotation (DocumentationClass=true, Documentation(info="<html>
+
+<p>The following DC/DC converter topologies are currently included in the PowerConverters library.</p>
+
+<ul>
+<li>SinglePhase dimmer with <a href=\"modelica://Modelica.Electrical.PowerConverters.ACAC.SinglePhaseTriac\">triac</a></li>
+<li>MultiPhase indcution machine soft starter with <a href=\"modelica://Modelica.Electrical.PowerConverters.ACAC.MultiPhaseTriac\">triac</a></li>
+</ul>
```
Comment (dietmarw) [inline Modelica/Electrical/PowerConverters.mo]: Some typos and the use of Multiphase should be fixed.

Comment (dietmarw) [reply Modelica/Electrical/PowerConverters.mo]: Probably best to make a search for multi

**Comment by christiankral** [comment]
OK, changed MutliPhase to Polyphase and SinglePhase to Single-phase. These changes apply to the newly introduced components only. Existing mutli and single phase names and components will be changed in a different ticket and PR.

**Comment by christiankral** [comment]
The correct spellings are single-phase and polyphase. From this lingual spelling I derive the Modelica spelling `SinglePhase` and `Polyphase`. Correct?

File: Modelica/Electrical/PowerConverters.mo (hunk 1 above)
Comment (dietmarw) [reply Modelica/Electrical/PowerConverters.mo]: Still typo in "indcution machine"

**Comment by dietmarw** [comment]
Yes. In order to have some kind of persistent linkage between the camel case and the original spelling that is how I would write it. You should go over the changes one more time and keep an eye out for this.

File: Modelica/Electrical/PowerConverters.mo [hunk 2]
```diff
@@ -90,6 +92,41 @@ is provided.
...
+</p>
+<p>
+The <a href=\"modelica://Modelica.Electrical.PowerConverters.ACAC.Control.VoltageToAngle\">Voltage2Angle block</a> 
+calculates phase angle from reference voltage.
+</p>
+<p>
+To control the soft start of an induction machine, 
+the <a href=\"modelica://Modelica.Electrical.PowerConverters.ACAC.Control.SoftStartControl\">SoftStartControl block</a> 
+is provided. It applies a voltage ramp during start, setting the ramp on hold whenever the measured current exceeds the maximum current. 
+Furthermore, a ramp down can be applied for stoppng the drive.
```
Comment (beutlich) [inline Modelica/Electrical/PowerConverters.mo]:
```suggestion
Furthermore, a ramp down can be applied for stopping the drive.
```

File: Modelica/Electrical/PowerConverters.mo [hunk 3]
```diff
@@ -6943,6 +7574,460 @@ General information about DC/DC converters can be found at the
...
+          der(vRef) = -1/tRampDown;
+        else
+          der(vRef) = 0;
+        end if;
+        annotation (Documentation(info="<html>
+<p>
+This block models the functionality of a soft starter controller, controlling the output <code>vRef</code> in the range [0,1] with respect to nominal voltage.
+</p>
+<p>
+Boolean input <code>start = true</code> causes the output <code>vRef</code> to be rised according to a ramp: <code>vRef = vStart + (1 - vStart)*(time - t0)/tRampUp</code>.
```
Comment (beutlich) [inline Modelica/Electrical/PowerConverters.mo]:
```suggestion
Boolean input <code>start = true</code> causes the output <code>vRef</code> to be risen according to a ramp: <code>vRef = vStart + (1 - vStart)*(time - t0)/tRampUp</code>.
```

File: Modelica/Electrical/PowerConverters.mo [hunk 4]
```diff
@@ -6562,9 +7110,92 @@ The firing signal is generated by comparing the sampled duty cycle input with a
                 fillPattern=FillPattern.Solid), Text(extent={{
                     -150,-120},{150,-160}}, textString = "%name", lineColor = {0, 0, 255})}),
           Documentation(info="<html>
+<p>
 This model linearly transforms the input voltage signal into a duty cycle. For the unipolar case the input voltage range is between zero and <code>vMax</code>. In case of bipolar input the input voltage is in the range between <code>-vMax</code> and <code>vMax</code>.
+</p>
+<p>
+Note: This block is replaced by the improved <a href=\"modelica://Modelica.Electrical.PowerConverters.DCDC.Control.Voltage2DutyCycle\">Voltage2DutyCycleBlock block</a>.
```
Comment (beutlich) [inline Modelica/Electrical/PowerConverters.mo]:
```suggestion
Note: This block is replaced by the improved <a href=\"modelica://Modelica.Electrical.PowerConverters.DCDC.Control.Voltage2DutyCycle\">Voltage2DutyCycle block</a>.
```

File: Modelica/Electrical/PowerConverters.mo (hunk 2 above)
Comment (christiankral) [reply Modelica/Electrical/PowerConverters.mo]: GitHub claims that I have no permission to push to this repository... 

image

So I klick on `Resolve conversation`. I hope this accepts the proposed changes

Comment (christiankral) [reply Modelica/Electrical/PowerConverters.mo]: No, the resolved conversation seems to imply that I have to resolve the issue manually. OK.

Comment (beutlich) [reply Modelica/Electrical/PowerConverters.mo]: Only the creator of the PR (@AHaumer) can accept those code suggestions.

Comment (christiankral) [reply Modelica/Electrical/PowerConverters.mo]: OK. I then way for the push of @AHaumer.

Comment (AHaumer) [reply Modelica/Electrical/PowerConverters.mo]: done

File: Modelica/Electrical/PowerConverters.mo (hunk 4 above)
Comment (beutlich) [reply Modelica/Electrical/PowerConverters.mo]: Still not fixed.

**Comment by beutlich** [comment]
> 
> 
> Modelica.Electrical.PowerConverters.DCDC.Control.VoltageToDutyCycle misses the obsolete annotation;
> 
> ```modelica
> obsolete = "Obsolete block - use Modelica.Electrical.PowerConverters.DCDC.Control.Voltage2DutyCycle instead");
> ```

Please add it.

**Comment by beutlich** [comment]
Any chance we can get rid of the obsolete block VoltageToDutyCycle (by conversion to Voltage2DutyCycle)? Otherwise I am afraid of having VoltageToDutyCycle in the MSL v4.x.y far too long.

**Comment by christiankral** [comment]
We can just remove it. 

* `VoltageToDutyCycle` works with blocks
* `Voltage2DutyCycle` is based on equations

Therefore, the two blocks are not compatible in terms of their internally used blocks and equations. Yet, `Voltage2DutyCycle` is 100% compatible to `VoltageToDutyCycle` with respect to its operational behavior. 

As `Voltage2DutyCycle` is an extended version of `VoltageToDutyCycle` personally I have no problem to just remove `VoltageToDutyCycle`. In this case I would even keep the "old name" `VoltageToDutyCycle` which would then not require any conversion script.

**Comment by beutlich** [comment]
> In this case I would even keep the "old name" `VoltageToDutyCycle` which would then not require any conversion script.

This isn't possible, since the parameters are renamed: vMax -> vLim, useConstantMaximumVoltage -> useConstantVoltageLimit.

Furthermore, would be nice if output variable `v` could get a `unit="V"` attribute.

**Comment by beutlich** [comment]
I'd propose to move VoltageToDutyCycle to ObsoleteModelica4.

**Comment by beutlich** [comment]
You would need to rebase your branch on master or merge the master branch into your branch, since ObsoleteModelica4.mo already is available in master.

**Comment by christiankral** [comment]
I hope the rebase worked out somehow...

**Comment by christiankral** [comment]
@beutlich So if we move `VoltageToDutyCycle` to `ObsoleteModelica4` we do NOT need a conversion script, right?

**Comment by beutlich** [comment]
> 
> 
> @beutlich So if we move `VoltageToDutyCycle` to `ObsoleteModelica4` we do NOT need a conversion script, right?

Just the convertClass from MSL to ObsoleteModelica4.

**Comment by christiankral** [comment]
> Furthermore, would be nice if output variable `v` could get a `unit="V"` attribute.
This reminds me of #2494 which was closed... I will yet add the unit

**Comment by christiankral** [comment]
> > @beutlich So if we move `VoltageToDutyCycle` to `ObsoleteModelica4` we do NOT need a conversion script, right?
> 
> Just the convertClass from MSL to ObsoleteModelica4.

Yes, this one indeed.

**Comment by beutlich** [comment]
Shall I do it and resolve the merge conflict?

**Comment by christiankral** [comment]
Please wait a couple more moments...

**Comment by beutlich** [comment]
Oh dear.

**Comment by christiankral** [comment]
Did I do something (or everything) wrong :-(

**Comment by beutlich** [comment]
Does not like like a rebased or merged branch. Did you cherry-pick instead?

**Comment by christiankral** [comment]
No I used rebase in Gitkraken

**Comment by beutlich** [comment]
Hm. There should not be 112 commits then.

**Comment by beutlich** [comment]
Normally, if you rebase, you then need to push --force to AHaumer:PowerConverters to rewrite the history.

**Comment by christiankral** [comment]
Can you revert the crazy changes. I can then re-implement the good changes.

**Comment by beutlich** [comment]
Hang on. Doing it now.

**Comment by beutlich** [comment]
Done. You should now do: git reset --hard AHaumer/PowerConverters

**Comment by christiankral** [comment]
Thanks a lot for cleaning up branch.

**Comment by christiankral** [comment]
Does that also mean that everything is done?

**Comment by beutlich** [comment]
A conversion testmodel might be straight-forward to add.

**Comment by christiankral** [comment]
Yes, I will provide it.

**Comment by christiankral** [comment]
OK, I guess everything is done.

**Comment by beutlich** [comment]
> 
> 
> OK, I guess everything is done.

Probably yes. But since I cannot judge on the behavioral changes I'll leave the approval to the experts.

**Comment by christiankral** [comment]
I added a test case to Modelica `Issue2970` such that the line

```modelica
toBeConverted = signal2mPulse.negativeEqual[1].y;
```

gets converted through this script statement: 

```modelica
convertElement("Modelica.Electrical.PowerConverters.ACDC.Control.Signal2mPulse",
               {"negativeEqual"},
               {"greaterNegative"});
```

Could that be that `negativeEqual` is a vector, which causes the conversion script not to work? @HansOlsson @beutlich Do you have an idea what I am making wrong?

**Comment by beutlich** [comment]
>  causes the conversion script not to work?

Cannot reproduce. Conversion successfully runs in Dymola 2020.

**Comment by christiankral** [comment]
Hmm..., I tested the conversion script on Dymola 2019 FD01.

**Comment by christiankral** [comment]
I consider the conversion script then to be valid.

**Comment by beutlich** [comment]
Also works in Dymola 2019 FD01 for me.

**Comment by christiankral** [comment]
The conversion script has not been working before for whatever reason. However, it now works in Dymola 2019 FD01 on my machine. Thanks for following the issue of the conversion script.

**Pull request merged**